src/bootstrap/native.rs

```diff
@@ -389,12 +393,82 @@ impl Step for Llvm {
 
         cfg.build();
 
+        if builder.config.bolt_profile_generate {
+            instrument_llvm_with_bolt(builder);
+        }
+        if let Some(path) = &builder.config.bolt_profile_use {
+            apply_bolt_profile_to_llvm(path);
+        }
+
         t!(stamp.write());
 
         build_llvm_config
     }
 }
 
+fn apply_bolt_profile_to_llvm(profile_path: &str) {
+    let lib_dir = PathBuf::from("/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/lib");
+    let llvm_lib_path = lib_dir.join("libLLVM-14-rust-1.61.0-nightly.so");
+    // The library cannot be instrumented in-place
+    let llvm_optimized_lib_path = PathBuf::from("/tmp/libLLVM-optimized.so");
+
+    let status = Command::new("llvm-bolt")
+        .arg(&llvm_lib_path)
+        .arg("-data")
+        .arg(&profile_path)
+        .arg("-o")
+        .arg(&llvm_optimized_lib_path)
+        .arg("-reorder-blocks")
+        .arg("cache+")
+        .arg("-reorder-functions")
+        .arg("hfsort")
+        .arg("-split-functions")
+        .arg("2")
+        .arg("-split-all-cold")
+        .arg("-split-eh")
+        .arg("-dyno-stats")
+        .status()
+        .expect("Could not instrument LLVM library using BOLT");
+
+    eprintln!("BOLT status: {}", status);
+
+    std::fs::copy(&llvm_optimized_lib_path, llvm_lib_path)
+        .expect("Cannot copy optimized LLVM library");
+    std::fs::remove_file(llvm_optimized_lib_path).expect("Cannot delete optimized LLVM library");
+}
+
+fn instrument_llvm_with_bolt(_builder: &Builder<'_>) {
+    // let lib_dir = builder.llvm_out(builder.config.build).join("lib");
+    // temporary hack
+    let lib_dir = PathBuf::from("/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/lib");
```

Comment:
The proper way to get this path would be to use `builder.llvm_out(builder.config.build).join("lib")`. However, even though this is the proper directory where LLVM is installed, when LLVM is finally packaged into the `stageN/lib` directory, it is actually being copied from `llvm/build/lib` (the build directory) and not from `llvm/lib` (the install directory).

The packaging code uses `llvm-config` to get the path of the LLVM lib to copy. The `llvm-config` is taken from the `llvm/build/bin` directory because of [this](https://github.com/rust-lang/rust/blob/master/src/bootstrap/native.rs#L79) line. @Mark-Simulacrum how should we solve this? Duplicate the `!msvc || is_ninja` check here? Or use `prebuilt_llvm_config` immediately after we build LLVM to get the path of the LLVM lib to instrument?

Reply:
ensure(Llvm { ... }) should return the llvm-config path, which you can then use to get the appropriate file path, I think?

Not quite sure if that helps answer the question.